src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs

```diff
@@ -34,34 +34,36 @@ internal sealed partial class AssetProvider
+            using var _1 = PooledDictionary<Checksum, object>.GetInstance(out var checksumToObjects);
+
+            SolutionStateChecksums stateChecksums;
             using (await s_gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
             {
-                // this will make 4 round trip to data source (VS) to get all assets that belong to the given solution checksum
+                // first, get top level solution state for the given solution checksum
+                var compilationStateChecksums = await _assetProvider.GetAssetAsync<SolutionCompilationStateChecksums>(
+                    assetHint: AssetHint.SolutionOnly, solutionChecksum, cancellationToken).ConfigureAwait(false);
 
-                // first, get solution checksum object for the given solution checksum
-                var solutionCompilationChecksumObject = await _assetProvider.GetAssetAsync<SolutionCompilationStateChecksums>(
-                    assetHint: AssetHint.None, solutionChecksum, cancellationToken).ConfigureAwait(false);
-                solutionChecksumObject = await _assetProvider.GetAssetAsync<SolutionStateChecksums>(
-                    assetHint: AssetHint.None, solutionCompilationChecksumObject.SolutionState, cancellationToken).ConfigureAwait(false);
+                using var _2 = PooledHashSet<Checksum>.GetInstance(out var checksums);
 
-                // second, get direct children of the solution
-                {
-                    using var _ = PooledHashSet<Checksum>.GetInstance(out var checksums);
+                // second, get direct children of the solution compilation state.
+                compilationStateChecksums.AddAllTo(checksums);
+                await _assetProvider.SynchronizeAssetsAsync(assetHint: AssetHint.SolutionOnly, checksums, results: null, cancellationToken).ConfigureAwait(false);
 
-                    checksums.Add(solutionCompilationChecksumObject.SourceGeneratorExecutionVersionMap);
-                    solutionChecksumObject.AddAllTo(checksums);
-                    await _assetProvider.SynchronizeAssetsAsync(assetHint: AssetHint.None, checksums, results: null, cancellationToken).ConfigureAwait(false);
-                }
+                // third, get direct children of the solution state.
+                stateChecksums = await _assetProvider.GetAssetAsync<SolutionStateChecksums>(
+                    assetHint: AssetHint.SolutionOnly, compilationStateChecksums.SolutionState, cancellationToken).ConfigureAwait(false);
+
+                // Ask for solutions and top-level projects as the solution checksums will contain the checksums for
+                // the project states and we want to get that all in one batch.
+                checksums.Clear();
+                stateChecksums.AddAllTo(checksums);
+                await _assetProvider.SynchronizeAssetsAsync(assetHint: AssetHint.SolutionAndTopLevelProjectsOnly, checksums, checksumToObjects, cancellationToken).ConfigureAwait(false);
```

Comment:
this is one of the places that uses SolutionAndTopLevelProjectsOnly as it is asking for all the ProjectStateChecksums for all projects in bulk.  I'm skeptical we need this (as opposed to just querying for them one at a time in the loop below.  But i'm going to test that in another PR.